bot/cogs/moderation/utils.py

```diff
@@ -74,22 +107,21 @@ def proxy_user(user_id: str) -> discord.Object:
     if expires_at:
         payload['expires_at'] = expires_at.isoformat()
 
-    try:
-        response = await ctx.bot.api_client.post('bot/infractions', json=payload)
-    except ResponseCodeError as exp:
-        if exp.status == 400 and 'user' in exp.response_json:
-            log.info(
-                f"{ctx.author} tried to add a {infr_type} infraction to `{user.id}`, "
-                "but that user id was not found in the database."
-            )
-            await ctx.send(
-                f":x: Cannot add infraction, the specified user is not known to the database."
-            )
-            return
+    # Try to apply the infraction. If it fails because the user doesn't exist, try to add it.
+    for attempt in range(2):
+        try:
+            response = await ctx.bot.api_client.post('bot/infractions', json=payload)
+        except ResponseCodeError as exp:
+            if exp.status == 400 and 'user'in exp.response_json:
+                # Only once attempt to try to add the user to the database, not two:
+                if attempt > 0 or await post_user(ctx, user) is None:
+                    return
+            else:
+                log.exception("An unexpected ResponseCodeError occurred while adding an infraction:")
+                await ctx.send(":x: There was an error adding the infraction.")
+                return
```

Comment:
This loop makes it a bit difficult to understand what's going on. It should be equivalent to something like this:
```
  try:
      response = ...post infraction
  except ...:
      if exception == user doesn't exist:
          post_user()
          try:
              response = ...post infraction
          except ...:
              log.trace(...)
          return
      else:
          log.trace(...)
          return
```